core/src/main/java/bisq/core/dao/node/parser/BlockParser.java

```diff
@@ -106,14 +104,13 @@ public Block parseBlock(RawBlock rawBlock) throws BlockHashNotConnectingExceptio
         // one get resolved.
         // Lately there is a patter with 24 iterations observed
         long startTs = System.currentTimeMillis();
-        List<Tx> txList = block.getTxs();
 
         rawBlock.getRawTxs().forEach(rawTx ->
                 txParser.findTx(rawTx,
                         genesisTxId,
                         genesisBlockHeight,
                         genesisTotalSupply)
-                        .ifPresent(txList::add));
+                        .ifPresent(daoStateService::onNewTxForLastBlock));
```

Comment:
I didn't want anything outside of DaoStateService to be modifying the block list or the tx list within each block, since the txMap cache/index must be kept in sync at all times. In particular, the block parser is calling TxParser.findTx before each new tx is added, which in turn calls DaoStateService, presumably relying on DaoState being consistent. I think this is why it's difficult to make Block.txs completely immutable, since tx's in the partially constructed last block can depend upon each other.

Reply:
Ok, I read the comment above this `forEach` and followed the `TxInputParser` code. There is definitely a non-encapsulated relationship between the DaoStateService and Block.

In order to not make it worse, I might suggest leaving this as `block.getTxns().add(txn)` and having the `onParseBlockComplete()` update the cache, instead.

This fits more of the existing model where the `BlockParser` creates the object and the `DaoStateService` takes that object and uses it. It seems natural that the `DaoStateService::onParseBlockComplete` could be used to update cached data that it needs for its public API.